packages/coordinator/src/services/communication/mtls-spine-client.ts

```diff
@@ -0,0 +1,197 @@
+    const {address, body, headers} = this.prepareSpineRequest(req)
+
+    try {
+      logger.info(`Attempting to send message to ${address}`)
+
+      const response = await axios.post<string>(
+        address,
+        body,
+        {
+          headers: headers,
+          httpsAgent: this.httpsAgent
+        }
+      )
+      return MtlsSpineClient.handlePollableOrImmediateResponse(response, logger)
+    } catch (error) {
+      logger.error(`Failed post request for spine client send. Error: ${error}`)
+      return MtlsSpineClient.handleError(error)
+    }
+  }
+
+  async poll(path: string, fromAsid: string, logger: pino.Logger): Promise<spine.SpineResponse<unknown>> {
+    const address = this.getSpineUrlForPolling(path)
+
+    logger.info(`Attempting to send polling message to ${address}`)
+
+    try {
+      const result = await axios.get<string>(
+        address,
+        {
+          headers: {
+            "nhsd-asid": fromAsid
+          },
+          httpsAgent: this.httpsAgent
+        }
+      )
+      return MtlsSpineClient.handlePollableOrImmediateResponse(result, logger, `/_poll/${path}`)
+    } catch (error) {
+      logger.error(`Failed polling request for polling path ${path}. Error: ${error}`)
+      return MtlsSpineClient.handleError(error)
+    }
+  }
+
+  private static handlePollableOrImmediateResponse(
+    result: AxiosResponse,
+    logger: pino.Logger,
+    previousPollingUrl?: string
+  ) {
+    if (result.status === 200) {
+      return this.handleImmediateResponse(result, logger)
+    }
+
+    if (result.status === 202) {
+      logger.info("Successful request, returning SpinePollableResponse")
+      const contentLocation = result.headers["content-location"]
+      const relativePollingUrl = contentLocation ? contentLocation : previousPollingUrl
+      logger.info(`Got content location ${contentLocation}. Using polling URL ${relativePollingUrl}`)
+      return {
+        pollingUrl: `${BASE_PATH}${relativePollingUrl}`,
+        statusCode: result.status
+      }
+    }
+
+    logger.error(`Got the following response from spine:\n${result.data}`)
+    throw Error(`Unsupported status, expected 200 or 202, got ${result.status}`)
+  }
+
+  private static handleImmediateResponse(
+    result: AxiosResponse,
+    logger: pino.Logger
+  ) {
+    logger.info("Successful request, returning SpineDirectResponse")
+    return {
+      body: result.data,
+      statusCode: result.status
+    }
+  }
+
+  private static handleError(error: Error): spine.SpineDirectResponse<unknown> {
+    const axiosError = error as AxiosError
+    if (axiosError.response) {
+      return {
+        body: axiosError.response.data,
+        statusCode: axiosError.response.status
+      }
+    } else {
+      return {
+        body: axiosError.message,
+        statusCode: 500
+      }
+    }
+  }
+
+  private getSpineEndpoint(requestPath?: string) {
+    return `${SPINE_URL_SCHEME}://${this.spineEndpoint}/${requestPath}`
```

Comment:
We should probably sanitise provided inputs, otherwise we may be open to injection attacks. It's an easy add though,
```typescript
private getSpineUrlForPolling(path: string) {
  const sanitizedPath = encodeURIComponent(path);
  return this.getSpineEndpoint(`_poll/${sanitizedPath}`);
}
```
I think should do it

Reply:
The encodeURIComponent has been implemented.